### PR TITLE
workspace/xreferences: optimizations; correct vendor handling; minor cleanups

### PR DESCRIPTION
**Pull request by slimsag**
Based on #88 (I will drop those commits once it is merged)

### CODE REVIEW

**Comment by keegancsmith** [comment]
Rebased for you, will review now.

File: langserver/workspace_refs.go [hunk 1]
```diff
@@ -70,27 +72,36 @@ func (h *LangHandler) handleWorkspaceReferences(ctx context.Context, conn JSONRP
...
-	// Collect dependency references.
-	results := refResultSorter{results: make([]lspext.ReferenceInformation, 0)}
-	par := parallel.NewRun(parallelism)
-	for _, pkg := range prog.Imported {
-		par.Acquire()
-		go func(pkg *loader.PackageInfo) {
-			defer par.Release()
+	// Collect dependency references in the AfterTypeCheck phase. This enables
+	// us to begin looking at packages as they are typechecked, instead of
+	// waiting for all packages to be typechecked (which is IO bound).
```
Comment (keegancsmith) [inline langserver/workspace_refs.go]: neat

File: langserver/workspace_refs.go [hunk 2]
```diff
@@ -70,27 +72,36 @@ func (h *LangHandler) handleWorkspaceReferences(ctx context.Context, conn JSONRP
...
+		started      uint32
+	)
+	afterTypeCheck := func(pkg *loader.PackageInfo, files []*ast.File) {
+		_, interested := unvendoredPackages[pkg.Pkg.Path()]
+		if !interested {
+			return
+		}
+
+		// Do not block the type-checker, but also avoid running with unbounded
+		// parallelism.
```
Comment (keegancsmith) [inline langserver/workspace_refs.go]: given typechecking is IO bound, surely this function is faster? Can't we just do unbounded parallelism to simplify?

File: langserver/workspace_refs.go [hunk 3]
```diff
@@ -100,19 +111,29 @@ func (h *LangHandler) handleWorkspaceReferences(ctx context.Context, conn JSONRP
...
+		}()
+	}
+
+	_, err := h.workspaceRefsTypecheck(ctx, bctx, conn, fset, pkgs, afterTypeCheck)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait for all worker goroutines to complete.
+	for i := 0; i < int(started); i++ {
```
Comment (keegancsmith) [inline langserver/workspace_refs.go]: technically you should read started via the atomic pkg. I do realise though that when we get here, there should be no concurrent accesses going on.

Also it seems your started + finishSignal could be accomplished with just a `sync.WaitGroup`

**Comment by slimsag** [comment]
Thanks for the rebase & review! I've used the `sync.WaitGroup` as you suggested, I knew I was missing something there (seemed oddly difficult to do).

**Pull request merged**